k8s.go

```diff
@@ -0,0 +1,643 @@
+
+	return nil
+}
+
+// etcdVarClient implements the libcalico-go backend api.Client interface.  It is used to
+// write the KVPairs retrieved from the Kubernetes datastore driver into the KV map
+// using etcdv2 naming scheme.
+type etcdVarClient struct {
+	vars map[string]string
+}
+
+func (c *etcdVarClient) Create(kvp *model.KVPair) (*model.KVPair, error) {
+	log.Fatal("Create should not be invoked")
+	return nil, nil
+}
+
+func (c *etcdVarClient) Update(kvp *model.KVPair) (*model.KVPair, error) {
+	log.Fatal("Update should not be invoked")
+	return nil, nil
+}
+
+func (c *etcdVarClient) Apply(kvp *model.KVPair) (*model.KVPair, error) {
+	path, err := model.KeyToDefaultPath(kvp.Key)
+	if err != nil {
+		log.Error("Unable to create path from Key: %s", kvp.Key)
+		return nil, err
+	}
+	value, err := model.SerializeValue(kvp)
+	if err != nil {
+		log.Error("Unable to serialize value: %s", kvp.Key)
+		return nil, err
+	}
+	c.vars[path] = string(value)
+	return kvp, nil
+}
+
+func (c *etcdVarClient) Delete(kvp *model.KVPair) error {
+	// Delete may be invoked as part of the Apply processing for  multi-key resource.
+	// However, since we start from an empty map each time, we never need to delete entries,
+	// so just ignore this request.
+	log.Debug("Delete ignored")
+	return nil
+}
+
+func (c *etcdVarClient) Get(key model.Key) (*model.KVPair, error) {
+	log.Fatal("Get should not be invoked")
+	return nil, nil
+}
+
+func (c *etcdVarClient) List(list model.ListInterface) ([]*model.KVPair, error) {
+	log.Fatal("List should not be invoked")
+	return nil, nil
+}
+
+func (c *etcdVarClient) Syncer(callbacks backendapi.SyncerCallbacks) backendapi.Syncer {
+	log.Fatal("Syncer should not be invoked")
+	return nil
+}
+
+func (c *etcdVarClient) EnsureInitialized() error {
+	log.Fatal("EnsureIntialized should not be invoked")
+	return nil
+}
+
+func (c *etcdVarClient) EnsureCalicoNodeInitialized(node string) error {
+	log.Fatal("EnsureCalicoNodeInitialized should not be invoked")
+	return nil
+}
+
+type ipamCacheK8s struct {
+	mu            sync.RWMutex
+	m             map[string]*ipPool
+	server        *Server
+	updateHandler func(*ipPool) error
+	lastIPPool    map[string]string
+}
+
+// create new IPAM cache
+func NewIPAMCacheK8s(s *Server, updateHandler func(*ipPool) error) *ipamCacheK8s {
+	return &ipamCacheK8s{
+		m:             make(map[string]*ipPool),
+		server:        s,
+		updateHandler: updateHandler,
+	}
+}
+
+// match checks whether we have an IP pool which contains the given prefix.
+// If we have, it returns the pool.
+func (c *ipamCacheK8s) match(prefix string) *ipPool {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	for _, p := range c.m {
+		if p.contain(prefix) {
+			return p
+		}
+	}
+	return nil
+}
+
+// update updates the internal map with IPAM updates when the update
+// is new addtion to the map or changes the existing item, it calls
+// updateHandler
+func (c *ipamCacheK8s) update(ipPoolData interface{}, del bool) error {
```

Comment:
ipPoolData is always of string type. Do we need to use interface here?

Reply:
update is the implement of IpamCache interface type.